src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,306 @@
 package frc.robot.subsystems.vision;
 
 import edu.wpi.first.math.geometry.Pose2d;
-import edu.wpi.first.math.geometry.Rotation2d;
-import edu.wpi.first.util.datalog.DataLog;
-import edu.wpi.first.util.datalog.StringLogEntry;
-import edu.wpi.first.wpilibj.DataLogManager;
-import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.math.geometry.Translation2d;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 import frc.robot.Constants.VisionConstants;
-import frc.robot.Constants.FieldConstants;
 import frc.robot.extras.LimelightHelpers;
-import frc.robot.extras.LimelightHelpers.LimelightResults;
+import frc.robot.extras.LimelightHelpers.PoseEstimate;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 public class VisionSubsystem extends SubsystemBase {
 
-  private LimelightResults currentlyUsedLimelightResults = LimelightHelpers.getLatestResults(VisionConstants.SHOOTER_LIMELIGHT_NAME);
-  private String currentlyUsedLimelight = VisionConstants.SHOOTER_LIMELIGHT_NAME;
   private Pose2d lastSeenPose = new Pose2d();
-  private boolean isTeleop = false;
+  private double headingDegrees = 0;
+  private double headingRateDegreesPerSecond = 0;
+  private final Map<Integer, AtomicBoolean> runningThreads = new ConcurrentHashMap<>();
+  private ExecutorService executorService = Executors.newFixedThreadPool(3); // Adjust the pool size as needed
+
+  /**
+   * The pose estimates from the limelights in the following order {shooterLimelight,
+   * frontLeftLimelight, frontRightLimelight}
+   */
+  private PoseEstimate[] limelightEstimates;
 
   public VisionSubsystem() {
+    limelightEstimates = new PoseEstimate[3]; 
+    
+    for (int limelightNumber = 0; limelightNumber < limelightEstimates.length; limelightNumber++) {
```

Comment:
I'm not really sure what best practice is in this case. When we only have 3 limelights, a for loop feels a little unnecessary, but I guess hardcoding it for all 3 also seems a little bad.

I would personally say that because we seem to type out all 3 other place in the code, we should do it here too, but what do you think?

Reply:
the reason i was trying this here was to see if it would help with garbage collection for threading (apparantly it’s slightly more memory efficient?? dont really know y), but the change thru VisualVM was negligible. ill prolly change it back

Reply:
Personally, I like the flexibility of the for loop, because it would be less code we have to change if our number of limelights changed. In theory, probably with changes in other places, we could be resilient against losing a limelight in such a way that our performance would get better if we detect that we lost a limelight. whether this is actually ever helpful is a different question, because we should have designed around our performance limitations (if that makes sense, if not I can explain more irl.)